config/config.go

```diff
@@ -136,37 +137,72 @@ const (
 )
 
 type Metrics struct {
-	Type   MetricsType `mapstructure:"type"`
-	Influx Influx      `mapstructure:"influx"`
+	Influx     InfluxMetrics     `mapstructure:"influx"`
+	Prometheus PrometheusMetrics `mapstructure:"prometheus"`
 }
 
 func (cfg *Metrics) validateAndLog() {
-	log.Infof("config.metrics.type: %s", cfg.Type)
-	switch cfg.Type {
-	case MetricsNone:
-	case MetricsInflux:
-		cfg.Influx.validateAndLog()
-	default:
-		log.Fatalf(`invalid config.metrics.type: %s. It must be "none" or "influx"`, cfg.Type)
+	influxEnabled := cfg.Influx.Enabled && cfg.Influx.validateAndLogMetricsData()
+	prometheusEnabled := cfg.Prometheus.Enabled && cfg.Prometheus.validateAndLogMetricsData()
+
+	if !influxEnabled && !prometheusEnabled {
+		log.Fatalf(`No metrics correctly specified in configuration file`)
 	}
 }
 
-type MetricsType string
-
-const (
-	MetricsNone   MetricsType = "none"
-	MetricsInflux MetricsType = "influx"
-)
-
-type Influx struct {
+type InfluxMetrics struct {
 	Host     string `mapstructure:"host"`
 	Database string `mapstructure:"database"`
 	Username string `mapstructure:"username"`
 	Password string `mapstructure:"password"`
+	Enabled  bool   `mapstructure:"enabled"`
+}
+
+func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool {
+	if influxMetricsConfig.Host != "" {
+		influxMetricsConfig.Enabled = false
+		log.Fatalf(`Despite being enabled, influx metrics came with no host info: config.metrics.influx.host = "".`)
+	}
+	if influxMetricsConfig.Database != "" {
+		influxMetricsConfig.Enabled = false
+		log.Fatalf(`Despite being enabled, influx metrics came with no database info: config.metrics.influx.database = "".`)
+	}
+	if influxMetricsConfig.Enabled {
+		log.Infof("config.metrics.influx.host: %s", influxMetricsConfig.Host)
+		log.Infof("config.metrics.influx.database: %s", influxMetricsConfig.Database)
+	}
+	return influxMetricsConfig.Enabled
+}
+
+type PrometheusMetrics struct {
+	Port             int    `mapstructure:"port"`
+	Namespace        string `mapstructure:"namespace"`
+	Subsystem        string `mapstructure:"subsystem"`
+	TimeoutMillisRaw int    `mapstructure:"timeout_ms"`
+	Enabled          bool   `mapstructure:"enabled"`
+}
+
+func (promMetricsConfig *PrometheusMetrics) validateAndLogMetricsData() bool {
+	if promMetricsConfig.Port == 0 {
+		promMetricsConfig.Enabled = false
+		log.Fatalf(`Despite being enabled, prometheus metrics came with an empty port number: config.metrics.prometheus.port = 0`)
+	}
+	if promMetricsConfig.Namespace != "" {
```

Comment:
This comparison is incorrect. Should be `promMetricsConfig.Namespace == ""`. I'm guessing the logic was reversed in an earlier draft.

Reply:
corrected